File: spikes/python/src/tt/platform/action.py
"""What an action is, with no transport in sight.

Nothing in this file mentions JSON. ``Payload`` is whatever Pydantic model the
action names; the JSON edge both frontends talk to is ``wire``, which adds the
schema and the decoder where they belong.

``execute`` is handed the open ``Session`` and writes for itself. There is no
second kind for creation: an ``INSERT`` and an ``UPDATE`` are both just things an
``execute`` body does with the session it holds, so what tells them apart is the
write each one issues, not the type each one returns.
"""

from dataclasses import dataclass
from typing import ClassVar

from pydantic import BaseModel
from sqlalchemy.orm import Session

from tt.platform.error import Conflict


@dataclass(frozen=True)
class Runnable:
    """The action applies and can run now."""


@dataclass(frozen=True)
class Refused:
    """The action applies but is refused right now, and says why.

    The reason is per-object, not a fixed literal: ``editStatus`` says "finish or
    drop 3 issues first" and the 3 is this project's count.
    """

    reason: str


# What an object offers for one action. ``None`` is "does not apply to this
# object at all — not offered", which is what lets ``wire.available`` drop the
# absent ones and keep the refused ones with their reason.
type Offered = Runnable | Refused
type Availability = Offered | None


class Checked:
    """Proof that availability was checked, threaded into ``execute``.

    In Rust this is a token whose constructor is private to the action module,
    and in OCaml an ``.mli`` hides ``execute`` outright — either way ``execute``
    is unreachable except through ``run``. Python can seal neither: this marker
    mirrors the shape, but nothing stops a caller building one and calling
    ``execute`` against an object that refused it. The enforcement point is a
    convention here rather than a type, and that gap is the Python finding.
    """

    __slots__ = ()


def _decide(available: bool, disabled: str | None) -> Availability:
    if not available:
        return None
    return Runnable() if disabled is None else Refused(disabled)


def _enforce(key: str, availability: Availability) -> Checked:
    """The enforcement point, as a value: a token comes back only when the object
    allows the write, and a refusal is raised otherwise."""
    match availability:
        case None:
            raise Conflict(f"{key} does not apply")
        case Refused(reason):
            raise Conflict(reason)
        case Runnable():
            return Checked()


class Action[Obj, P: BaseModel]:
    """One action, declared in one place.

    A subclass sets ``KEY`` and ``Payload``, overrides whichever of the two hooks
    it needs, and writes ``execute``. ``availability`` and ``run`` come from here:
    ``run`` is the one path to ``execute``, and it enforces availability against
    the object first.
    """

    KEY: ClassVar[str]
    # The Pydantic model of the arguments, from which ``wire`` derives both the
    # advertised schema and the decoder — so the two cannot name different types.
    Payload: ClassVar[type[BaseModel]]

    @classmethod
    def is_available(cls, obj: Obj) -> bool:
        """False when the action does not apply to this object at all."""
        return True

    @classmethod
    def is_disabled(cls, obj: Obj) -> str | None:
        """The reason it applies but cannot run right now, or ``None``."""
        return None

    @classmethod
    def execute(cls, obj: Obj, payload: P, tx: Session, checked: Checked) -> str:
        """Do the thing, on the open session, and return the message a frontend
        reports. An edit is an ``UPDATE``, a delete stamps ``deleted_at``, a
        creator ``INSERT``s a different table — and nothing outside the body says
        which."""
        raise NotImplementedError

    @classmethod
    def availability(cls, obj: Obj) -> Availability:
        return _decide(cls.is_available(obj), cls.is_disabled(obj))

    @classmethod
    def run(cls, obj: Obj, payload: P, tx: Session) -> str:
        """Availability is enforced here rather than at the edge, so a caller that
        already holds a payload cannot skip it. ``wire.dispatch`` decodes and then
        comes through here."""
        checked = _enforce(cls.KEY, cls.availability(obj))
        return cls.execute(obj, payload, tx, checked)

File: spikes/rust/src/platform/action.rs
//! What an action is, with no transport in sight.
//!
//! Nothing in this file mentions `serde` or `serde_json`. `Payload` carries no
//! bounds — an action is free to have a payload that is not serialisable at
//! all. The JSON edge both frontends talk to is [`crate::platform::wire`],
//! which adds the bounds where they belong.
//!
//! `execute` is handed the open transaction and writes for itself. There is no
//! second trait for creation: an `INSERT` and an `UPDATE` are both just things
//! an `execute` body does with the connection it holds, so what tells them
//! apart is the write each one issues, not the type each one returns.

use std::borrow::Cow;

use sea_orm::DatabaseTransaction;

use crate::platform::error::Error;

/// An action that applies to the object. `Runnable` cannot carry a reason and
/// `Refused` cannot lack one.
///
/// The reason is a [`Cow`] rather than a `&'static str` because a refusal is
/// allowed to quote the object it read: `editStatus` says "finish or drop 3
/// issues first", and the 3 is per-project. Every refusal that does not need
/// that stays a borrowed literal.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Offered {
    /// Applies, but is refused right now.
    Refused(Cow<'static, str>),
    /// Can run now.
    Runnable,
}

/// Whether one action applies to one object, and if so whether it can run.
///
/// `None` is "does not apply to this object at all — not offered". Naming it
/// `Option` is what lets [`crate::platform::wire::available`] return `Offered`,
/// so a caller of that function cannot be handed an absent action.
pub type Availability = Option<Offered>;

/// Proof that availability was checked. Its field is private, so no module but
/// this one can build one — which is what makes [`Action::run`] the only way to
/// reach [`Action::execute`].
///
/// It is the Rust half of the enforcement guarantee: a token that only
/// [`enforce`] mints, threaded into a body that cannot be reached without it.
pub struct Checked(());

/// The decision an action makes out of its two hooks.
fn decide(available: bool, disabled: Option<Cow<'static, str>>) -> Availability {
    if !available {
        None
    } else {
        Some(match disabled {
            Some(reason) => Offered::Refused(reason),
            None => Offered::Runnable,
        })
    }
}

/// The enforcement point, as a value: the token comes back only when
/// availability allows the write.
fn enforce(key: &str, availability: Availability) -> Result<Checked, Error> {
    match availability {
        None => Err(Error::Conflict(format!("{key} does not apply"))),
        Some(Offered::Refused(reason)) => Err(Error::Conflict(reason.into_owned())),
        Some(Offered::Runnable) => Ok(Checked(())),
    }
}

/// One action, declared in one place.
pub trait Action {
    type Obj;
    type Payload;

    const KEY: &str;

    /// False when the action does not apply to this object at all.
    fn is_available(_obj: &Self::Obj) -> bool {
        true
    }

    /// The reason it applies but cannot run right now.
    fn is_disabled(_obj: &Self::Obj) -> Option<Cow<'static, str>> {
        None
    }

    /// Do the thing, on the open transaction, and return the message a frontend
    /// reports. An edit is an `UPDATE`, a delete stamps `deleted_at`, a creator
    /// `INSERT`s a different table — and nothing outside the body says which.
    ///
    /// This is the one place a concrete [`DatabaseTransaction`] is named rather
    /// than a connection type parameter. OCaml keeps the connection opaque and
    /// threads it through, but Rust's boxed-future erasure (see
    /// [`crate::platform::wire`]) makes the generic form costly for no gain
    /// here, so the transaction is spelled out. A service query stays generic
    /// over the connection, so the body still runs the same read against the
    /// live database or against this transaction.
    fn execute(
        obj: Self::Obj,
        payload: Self::Payload,
        tx: &DatabaseTransaction,
        _: Checked,
    ) -> impl std::future::Future<Output = Result<String, Error>>;

    fn availability(obj: &Self::Obj) -> Availability {
        decide(Self::is_available(obj), Self::is_disabled(obj))
    }

    /// Availability is enforced here rather than at the edge, so a caller that
    /// already holds a payload cannot skip it. [`crate::platform::wire::dispatch`]
    /// decodes and then comes through this.
    fn run(
        obj: Self::Obj,
        payload: Self::Payload,
        tx: &DatabaseTransaction,
    ) -> impl std::future::Future<Output = Result<String, Error>> {
        async move {
            let checked = enforce(Self::KEY, Self::availability(&obj))?;
            Self::execute(obj, payload, tx, checked).await
        }
    }
}
